File: 049_GroupAnagrams.py

class Solution(object):
    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        """
        if not strs:
            return None

        set_in = {}
        for i in xrange(len(strs)):
            tmp =[]
            for j in xrange(len(strs[i])):
                tmp.append(strs[i][j])
            tmp.sort()
            key_s = ''.join(tmp)
            if not  set_in.has_key(key_s):
                str_l = []
                str_l.append(strs[i])
                set_in[key_s]=str_l
            else:
                str_l = set_in[key_s]
                str_l.append(strs[i])
                set_in.update({key_s:str_l})

        return  set_in.values()

if __name__ == '__main__':

    strs = ["eat", "tea", "tan", "ate", "nat", "bat"]
    k = Solution()
    print 'input :',strs
    r = k.groupAnagrams(strs)
    print  'result : ',r

File: 049_GroupAnagrams.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function(strs) {
    //Runtime: 140 ms, faster than 65.77% of JavaScript online submissions for Group Anagrams.
    //Memory Usage: 46.7 MB, less than 29.09% of JavaScript online submissions for Group Anagrams.
    var re = {};
    var i = 0;
    var tmp = [];
    for (i=0;i<strs.length;i++){
        tmp = strs[i].split('').sort().join('');
        if(re[tmp]===undefined){
            re[tmp]= [];
        };
        re[tmp].push(strs[i]);
    };  
    return(Object.keys(re).map(function(key){return(re[key]);}));    
};

/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams1 = function(strs) {
    //Runtime: 128 ms, faster than 89.24% of JavaScript online submissions for Group Anagrams.
    //Memory Usage: 44.3 MB, less than 96.36% of JavaScript online 
    var len=strs.length;
    var str=[];
    var res=new Map();
    for(var i=0;i<len;i++){
        //取键
        str=strs[i].split('').sort().join('');
        //存值
        if(!res.has(str)) 
            res.set(str,new Array())
        
        res.get(str).push(strs[i]);
    }
    return [...res.values()]
};

    

var str_in = ["eat", "tea", "tan", "ate", "nat", "bat"];
console.log(str_in);
console.log(groupAnagrams(str_in));
console.log('-------the end!--------');